### PR TITLE
Python Rewrite

### PR DESCRIPTION
**Pull request by murjax**
Rewrite the original Python project (currently in this repo under old/python).

Improvements made:

- Remove git library dependency. We just run `git clone` directly. This project no longer requires any dependencies.
- Remove container folder for repos. Ordinarily, we clone repos in the current working directory. Optimally, this executable is added to the consumer's shell path so it can be used in place of a standard `git clone` command.
- Add tests.

### CODE REVIEW

File: python/.gitignore [hunk 1]
```diff
@@ -0,0 +1,160 @@
+# Byte-compiled / optimized / DLL files
```
Comment (MichaelDimmitt) [inline python/.gitignore:1]: Not a big deal but why are there so many files tracked by the gitignore?

Comment (murjax) [reply python/.gitignore:1]: I just added a generic Python gitignore: https://github.com/github/gitignore/blob/main/Python.gitignore

File: python/github_repo_cloner/clone_handler.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    if username == None or username == '':
+        return
+
+    repositories = get_repositories(username)
+    if len(repositories) == 0:
+        return
+
+    commands = map(create_command, repositories)
+    final_command = ' & '.join(commands)
+    subprocess.run(final_command, shell=True, check=False)
```
Comment (MichaelDimmitt) [inline python/github_repo_cloner/clone_handler.py:14]: this does not look like it spawns more than one process.

File: python/.gitignore (hunk 1 above)
Comment (MichaelDimmitt) [reply python/.gitignore:1]: Cool, gitignores 1 level down enable us not to have tracking at the top level!

File: python/github_repo_cloner/clone_handler.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    """Github Clone Handler"""
+    PER_PAGE = 30.0
+    BASE_URL = 'https://api.github.com'
+
+    def __init__(self, username):
+        self.username = username
+        self.user_info = {}
+        self.repo_info = []
+        self.errors = []
+
```
Comment (MichaelDimmitt) [inline python/github_repo_cloner/clone_handler.py:16]: @murjax , how do you order your methods in a class?

1. Is there a logic to it like the items at the  bottom are executed before the ones at the top?
2. Or do you order alphabetically when there is no reason to have any order?

3. Or just write the function where ever when you have a need for it?

File: python/github_repo_cloner/clone_handler.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    @classmethod
+    def get_username(cls):
+        """Collect Github username"""
+        return input('Please enter your Github username: ')
```
Comment (MichaelDimmitt) [inline python/github_repo_cloner/clone_handler.py:20]: I think this probably should be in cloner.py

until it becomes repeatedly used enough to put it in an input class.

File: python/github_repo_cloner/clone_handler.py (hunk 3 above)
Comment (MichaelDimmitt) [reply python/github_repo_cloner/clone_handler.py:16]: Screenshot 2023-09-29 at 3 34 49 PM

File: python/github_repo_cloner/clone_handler.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        repo_count = self.user_info.get('public_repos', 0)
+        return math.ceil(repo_count / self.PER_PAGE)
+
+    def __get_repo_info_for_page(self, page):
+        response = requests.get(f'{self.BASE_URL}/users/{self.username}/repos?page={page}')
+        return response.json()
+
+    def __get_info(self):
+        self.__get_user_info()
+        self.__get_repo_info()
```
Comment (MichaelDimmitt) [inline python/github_repo_cloner/clone_handler.py:58]: nice!

File: python/github_repo_cloner/clone_handler.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+            self.__print_errors()
+            return False
+
+        subprocess.run(self.__final_command(), shell=True, check=False)
+        return True
+
+    # Private methods
+
+    def __final_command(self):
+        return ' & '.join(self.__clone_commands())
```
Comment (MichaelDimmitt) [inline python/github_repo_cloner/clone_handler.py:37]: consider:
__get_clone_commands()

**Review by MichaelDimmitt — APPROVED** [review]
added come comments but this has been held up long enough ship it!
address in a future pr if needed.

respond to comments in this pr even after merging it.

This also makes me wonder what to do for users that only have python 2.6 on their system.

but really we should probably be shipping binaries of all of these and not require them to have any of these languages installed.

in this regard we should have a /bin folder with all the versions.
let me know and I will make an issue about feature: add /bin folder.

File: python/github_repo_cloner/clone_handler.py (hunk 3 above)
Comment (murjax) [reply python/github_repo_cloner/clone_handler.py:16]: Mostly 3. I try to group methods close to where they are used. I don't follow it consistently though. Sometimes there are similar looking methods that are better grouped together but are used in different places.

File: python/github_repo_cloner/clone_handler.py (hunk 4 above)
Comment (murjax) [reply python/github_repo_cloner/clone_handler.py:20]: Yep, agreed. I'll move it.

**Comment by murjax** [comment]
>This also makes me wonder what to do for users that only have python 2.6 on their system.
>
>but really we should probably be shipping binaries of all of these and not require them to have any of these languages installed.
>
>in this regard we should have a /bin folder with all the versions.
>let me know and I will make an issue about feature: add /bin folder.

@MichaelDimmitt 
I'm not sure this is possible? Ruby binaries for instance can be run as executable but are still running Ruby under the hood. They are not compiled into machine code and therefore the end user must have Ruby installed. I think a similar case exists for Elixir. I couldn't figure out a Python binary yet.

I think we would have to write this in something that compiles into machine code such as Rust to make it portable.

**Pull request merged**